src/StreamJsonRpc.Tests/JsonRpcClient20InteropTests.cs

```diff
@@ -149,6 +150,147 @@ public async Task InvokeWithParameterPassedAsObjectAsync_NoParameter()
         Assert.Null(request["params"]);
     }
 
+    [Fact]
+    public async Task InvokeWithProgressParameterAsArray()
+    {
+        int sum = 0;
+        ProgressWithCompletion<int> progress = new ProgressWithCompletion<int>(report =>
+        {
+            sum += report;
+        });
+
+        int n = 3;
+        Task<int> invokeTask = this.clientRpc.InvokeAsync<int>("test", new object[] { n, progress });
+
+        JToken request = await this.ReceiveAsync();
+
+        long progressID = request["params"][1].Value<long>();
+
+        // Send responses as $/progress
+        int sum2 = 0;
+        for (int i = 0; i < n; i++)
+        {
+            string content = "{ \"jsonrpc\": \"2.0\", \"method\": \"$/progress\", \"params\": { \"token\": " + progressID + ", \"value\": " + i + ", } ,}";
+            JObject json = JObject.Parse(content);
+
+            this.Send(json);
+
+            sum2 += i;
+        }
+
+        System.Threading.Thread.Sleep(1000);
```

Comment:
>System.Threading.Thread.Sleep(1000); [](start = 8, length = 36)

Please never sleep in tests. If you had to insert a delay, use `await Task.Delay(1000);`. It allows concurrent test execution to go faster.
But in this case, it looks like you've got an unstable test since you're hoping that things happen within a second which on various machines and under different workloads will cause tests to non-deterministically fail. It also causes tests to run substantially slower than necessary since *usually* this will happen in a split second.
One thing you can do is have your client-side progress handler set an event every time it fires, that you await on. Then you can assert that each individual message is what you expect so we can for example verify that ordering is preserved as well.

Reply:
I'm unexperienced on how to use the event as a flag that I can await on to make sure the progress report is done, I investigated a little and I thought using ManualResetEvent might be the way, but I tried an approach using that (You can see the code of that approach commented) and it didn't really work.

---
In reply to: [303654071](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r303654071) [](ancestors = 303654071)

Reply:
Ya... we don't want to put these events into the product itself. Rather, we want them in the test code that responds to them. I'll conjure up or find you a sample.

---
In reply to: [304103873](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r304103873) [](ancestors = 304103873,303654071)

Reply:
Create an `AsyncAutoResetEvent` local in this test method. **Set** the event inside your progress callback. **await** the event where you currently have your WaitOne line.
Also when you await the event, add `.WithCancellation(this.TimeoutToken);` at the end so that if the event doesn't come in a timely way, the test will fail instead of hang.

---
In reply to: [304378299](https://github.com/milopezc/vs-streamjsonrpc/pull/1#discussion_r304378299) [](ancestors = 304378299,304103873,303654071)